modules/social_features/social_core/social_core.module

```diff
@@ -522,6 +524,35 @@ function social_core_filter_process_format($element) {
   return $element;
 }
 
+/**
+ * Hide blocks on extendable visibility path.
+ *
+ * Implements hook_block_access().
+ */
+function social_core_block_access(Block $block, $operation, AccountInterface $account) {
+  $blocks = &drupal_static(__FUNCTION__);
+  $blocks = [];
+
+  // Allow other modules to change the block visibility.
+  \Drupal::moduleHandler()->alter('social_core_block_visibility_path', $blocks);
```

Comment:
I like the optimisation of not running the hook more than once, but to be effective we should check if we actually need to recompute the blocks.

```suggestion
if (empty($blocks)) {
  \Drupal::moduleHandler()->alter('social_core_block_visibility_path', []);
 }
```

This also shows that you're not really altering anything but only collecting arrays. With this in mind it may be better to use https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21ModuleHandler.php/function/ModuleHandler%3A%3AinvokeAll/8.2.x

Note that that method already merges the arrays for you, so you may need to move to a slightly different array structure to ensure merging happens correctly. 
```
'block_id' => [
  'path' => TRUE,
]
```

The benefit of this is that Drupal takes care of merging the array for you and you can remove the loops from each of the hook implementations.

If you want to alter what another module does then you'd have to implement an alter hook separately. The added benefit of this is that page definitions can be altered mostly without changing the module weight (as they happen in separate hooks).

Moving the entire creation and caching of `$blocks` into a separate function can make it easier to reason about as well :)

Reply:
Moved to the invokeall.